ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -478,36 +490,17 @@ let check_operation_error ~__context ~vmr ~vmgmr ~ref ~clone_suspended_vm_enable
 
   current_error
 
-let maybe_get_guest_metrics ~__context ~ref =
-  if Db.is_valid_ref __context ref
-  then Some (Db.VM_guest_metrics.get_record_internal ~__context ~self:ref)
-  else None
-
-let get_info ~__context ~self =
-  let all = Db.VM.get_record_internal ~__context ~self in
-  let gm = maybe_get_guest_metrics ~__context ~ref:(all.Db_actions.vM_guest_metrics) in
-  let clone_suspended_vm_enabled = Helpers.clone_suspended_vm_enabled ~__context in
-  let vdis_reset_and_caching = List.filter_map (fun vbd ->
-      try
-        let vdi = Db.VBD.get_VDI ~__context ~self:vbd in
-        let sm_config = Db.VDI.get_sm_config ~__context ~self:vdi in
-        Some ((assoc_opt "on_boot" sm_config = Some "reset"), (bool_of_assoc "caching" sm_config))
-      with _ -> None) all.Db_actions.vM_VBDs in
-  all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching
-
 let get_operation_error ~__context ~self ~op ~strict =
-  let all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching = get_info ~__context ~self in
-  check_operation_error __context all gm self clone_suspended_vm_enabled vdis_reset_and_caching op strict
+  check_operation_error ~__context ~ref:self ~op ~strict
 
 let assert_operation_valid ~__context ~self ~op ~strict =
   match get_operation_error ~__context ~self ~op ~strict with
   | None       -> ()
   | Some (a,b) -> raise (Api_errors.Server_error (a,b))
 
 let update_allowed_operations ~__context ~self =
-  let all, gm, clone_suspended_vm_enabled, vdis_reset_and_caching = get_info ~__context ~self in
   let check accu op =
-    match check_operation_error __context all gm self clone_suspended_vm_enabled vdis_reset_and_caching op true with
```

Comment:
Realised that maybe we did things this way for performance reasons: get the info from the db only once, and use that info when we loop through multiple VM operations here, to avoid doing the same DB accesses multiple times. With this change we do a `Db.VM.get_record_internal`, `Db.VM_guest_metrics.get_record_internal`, and a `Db.VBD.get_VDI` & `Db.VDI.get_sm_config` for each `VBD` & `VDI`, during each iteration (however we could mostly eliminate the last two by only computing them when necessary or making them lazy). Not sure whether this is a performance issue, and how expensive/slow these DB accesses are.